### PR TITLE
[WIP] positional audio that updates? borrowed from goon sound2

### PR DESCRIPTION
**Pull request by Iamgoofball**
https://github.com/goonstation/goonstation/blob/master/code/modules/sound/~sound2.dm

## About The Pull Request

doesn't work fully yet, SOUND_UPDATE seems to be busted on tgcode and wont update

## Why It's Good For The Game

3d positional audio is good, actually

### CODE REVIEW

File: code/modules/mob/mob_movement.dm [hunk 1]
```diff
@@ -542,3 +542,10 @@
 /// Can this mob move between z levels
 /mob/proc/canZMove(direction, turf/target)
 	return FALSE
+
+/mob/Move()
```
Comment (optimumtact) [inline code/modules/mob/mob_movement.dm:546]: CAN YOU NOT, add it to the actual Move proc

File: code/game/atoms_movable.dm [hunk 2]
```diff
@@ -525,6 +525,12 @@
 
 	SEND_SIGNAL(src, COMSIG_MOVABLE_MOVED, OldLoc, Dir, Forced)
 
+	update_sounds()
+	for(var/mob/m in src)
+		if(m.client)
+			for(var/sound/s in GLOB.allAttachedSounds)
+				s.recalcClient(m.client)
+
```
Comment (LemonInTheDark) [inline code/game/atoms_movable.dm:533]: This seems like a lag spike waiting to happen

File: code/game/sound.dm [hunk 3]
```diff
@@ -180,6 +245,7 @@ distance_multiplier - Can be used to multiply the distance at which the sound is
 			S.echo[4] = 0 //RoomHF setting, 0 means normal reverb.
 
 	SEND_SOUND(src, S)
+	addtimer(CALLBACK(S, /sound.proc/stop), S.len + 10 SECONDS) // account for pitch shit
```
Comment (LemonInTheDark) [inline code/game/sound.dm:248]: I hate this, how often will it fire?

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
If we're taking this code from goonstation have you received permission from the authors to re-license it to agpl 3 or whatever?

File: code/game/sound.dm (hunk 3 above)
Comment (Iamgoofball) [reply code/game/sound.dm:248]: look bro unless sounds have a self delete this is what we gotta do

File: code/game/atoms_movable.dm [hunk 4]
```diff
@@ -525,6 +525,12 @@
 
 	SEND_SIGNAL(src, COMSIG_MOVABLE_MOVED, OldLoc, Dir, Forced)
 
+	update_sounds()
+	for(var/mob/m in src)
```
Comment (tralezab) [inline code/game/atoms_movable.dm:529]: better var name, please

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/game/atoms_movable.dm [hunk 5]
```diff
@@ -525,6 +525,12 @@
...
+		if(m.client)
+			for(var/sound/s in GLOB.allAttachedSounds)
```
Comment (tralezab) [inline code/game/atoms_movable.dm:531]: better var name, please

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/game/atoms_movable.dm [hunk 6]
```diff
@@ -1182,3 +1188,8 @@
 */
 /atom/movable/proc/keybind_face_direction(direction)
 	setDir(direction)
+
+/atom/movable/proc/update_sounds()
+	if(length(attachedSounds))
+		for (var/sound/S as anything in attachedSounds)
```
Comment (tralezab) [inline code/game/atoms_movable.dm:1194]: better var name, please

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/game/sound.dm [hunk 7]
```diff
@@ -22,6 +26,65 @@
 		0, // Flags (1 = Auto Direct, 2 = Auto Room, 4 = Auto RoomHF)
 	)
 	environment = SOUND_ENVIRONMENT_NONE //Default to none so sounds without overrides dont get reverb
+	var/atom/attached
+	var/list/listeners = list()
+	var/maxdist = 0
+
+/sound/proc/attach( var/atom/towhat )
```
Comment (tralezab) [inline code/game/sound.dm:33]: that var would be better in snake case, remove var/ part as that is unnecessary

also remove the spaces to the left and right of it, lol

File: code/game/sound.dm [hunk 8]
```diff
@@ -22,6 +26,65 @@
...
+/sound/proc/recalcClient(var/client/c)
+	if(!GLOB.mutecache)
+		GLOB.mutecache = sound(null)
+	GLOB.mutecache.channel = channel
+	status |= SOUND_UPDATE
+	var/turf/CT = get_turf(c.mob)
+	var/turf/AT = get_turf(attached)
+	var/dist = get_dist(CT, AT)
+	if( !listeners.Find(c) )
+		//world << "Their dist: [dist] vs [maxdist]"
```
Comment (tralezab) [inline code/game/sound.dm:66]: remove commented out code

File: code/game/sound.dm [hunk 9]
```diff
@@ -22,6 +26,65 @@
...
+			attached = null
+			listeners = list()
+
+/sound/proc/recalcClient(var/client/c)
+	if(!GLOB.mutecache)
+		GLOB.mutecache = sound(null)
+	GLOB.mutecache.channel = channel
+	status |= SOUND_UPDATE
+	var/turf/CT = get_turf(c.mob)
+	var/turf/AT = get_turf(attached)
```
Comment (tralezab) [inline code/game/sound.dm:63]: for both CT and AT:

better var name, please

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/game/sound.dm [hunk 10]
```diff
@@ -22,6 +26,65 @@
...
+			GLOB.mutecache = sound(null)
+		GLOB.mutecache.channel = channel
+		world << src
+		if(attached)
+			attached.attachedSounds.Remove( src )
+			GLOB.allAttachedSounds.Remove( src )
+			attached = null
+			listeners = list()
+
+/sound/proc/recalcClient(var/client/c)
```
Comment (tralezab) [inline code/game/sound.dm:57]: better var name, please

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/game/sound.dm [hunk 11]
```diff
@@ -22,6 +26,65 @@
...
+	else if( dist > maxdist )
+		listeners.Remove(c)
+		SEND_SOUND(c, GLOB.mutecache)
+	else
+		pan = (AT.x - CT.x)/maxdist*100
+		volume = 100-(dist/maxdist*100)
+		SEND_SOUND(c, src)
+
+/sound/proc/recalculate()
+	for(var/client/c in GLOB.clients)
```
Comment (tralezab) [inline code/game/sound.dm:84]: better var name, please

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/game/sound.dm [hunk 12]
```diff
@@ -22,6 +26,65 @@
...
+		//world << "Recalculating for [c]"
```
Comment (tralezab) [inline code/game/sound.dm:85]: remove commented out code

File: code/game/sound.dm [hunk 13]
```diff
@@ -81,11 +144,11 @@ falloff_distance - Distance at which falloff begins. Sound is at peak volume (in
 	for(var/P in listeners)
 		var/mob/M = P
 		if(get_dist(M, turf_source) <= maxdistance)
-			M.playsound_local(turf_source, soundin, vol, vary, frequency, falloff_exponent, channel, pressure_affected, S, maxdistance, falloff_distance, 1, use_reverb)
+			M.playsound_local(source, soundin, vol, vary, frequency, falloff_exponent, channel, pressure_affected, S, maxdistance, falloff_distance, 1, use_reverb)
 	for(var/P in SSmobs.dead_players_by_zlevel[source_z])
 		var/mob/M = P
 		if(get_dist(M, turf_source) <= maxdistance)
```
Comment (tralezab) [inline code/game/sound.dm:150]: if you're changing parts about this proc, you should really change all of these single letter vars to better ones

File: code/modules/mob/mob_movement.dm [hunk 14]
```diff
@@ -542,3 +542,10 @@
...
+	. = ..()
+	if(client)
+		for(var/sound/s in GLOB.allAttachedSounds)
```
Comment (tralezab) [inline code/modules/mob/mob_movement.dm:549]: better var name, please

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/modules/mob/mob_movement.dm [hunk 15]
```diff
@@ -542,3 +542,10 @@
...
+			s.recalcClient(client)
+			world << "recalculating for [src]"
```
Comment (tralezab) [inline code/modules/mob/mob_movement.dm:551]: uh... this is bad right? like, this should not be here, this is a debug print

File: code/game/sound.dm [hunk 16]
```diff
@@ -22,6 +26,65 @@
...
+		GLOB.channelReserved=256
+	if(!attached.attachedSounds)
+		attached.attachedSounds = list()
+	attached.attachedSounds.Add(src)
+	GLOB.allAttachedSounds.Add(src)
+
+	recalculate()
+
+/sound/proc/stop()
+	if(channel)
```
Comment (tralezab) [inline code/game/sound.dm:46]: early return please

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-early-return

File: code/game/atoms_movable.dm [hunk 17]
```diff
@@ -525,6 +525,12 @@
 
 	SEND_SIGNAL(src, COMSIG_MOVABLE_MOVED, OldLoc, Dir, Forced)
 
+	update_sounds()
+	for(var/mob/m in src)
+		if(m.client)
```
Comment (tralezab) [inline code/game/atoms_movable.dm:530]: early return please

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-early-return

File: code/game/atoms_movable.dm [hunk 18]
```diff
@@ -1182,3 +1188,8 @@
 */
 /atom/movable/proc/keybind_face_direction(direction)
 	setDir(direction)
+
+/atom/movable/proc/update_sounds()
+	if(length(attachedSounds))
```
Comment (tralezab) [inline code/game/atoms_movable.dm:1193]: early return please

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-early-return

File: code/modules/mob/mob_movement.dm [hunk 19]
```diff
@@ -542,3 +542,10 @@
 /// Can this mob move between z levels
 /mob/proc/canZMove(direction, turf/target)
 	return FALSE
+
+/mob/Move()
+	. = ..()
+	if(client)
```
Comment (tralezab) [inline code/modules/mob/mob_movement.dm:548]: early return please

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-early-return

**Review by tralezab — COMMENTED** [review]
here's some ways to bring this code up to contribution standards.

there was also some weirdly formatted code in this, `attach()` has some good examples:
``` dm
	if(GLOB.channelReserved>GLOB.channelMax)
		GLOB.channelReserved=256
```
no spaces? bleugh
``` dm
/sound/proc/attach( var/atom/towhat )
```
spaces before and after arguments? gross! they even do it in a block in `stop()`
``` dm
GLOB.allAttachedSounds.Remove( src )
```

anyway if maintainers care about that kind of stuff they'll tell you

File: code/game/sound.dm (hunk 3 above)
Comment (Mothblocks) [reply code/game/sound.dm:248]: We *cannot* be using timers like this--we have enough problems with them as is, adding htem one of the most called procs in the game is going to end in disaster.

What references are left to the sound after they're made and sent? It doesn't look like any.

File: code/game/sound.dm [hunk 20]
```diff
@@ -22,6 +26,65 @@
...
+	if(GLOB.channelReserved>GLOB.channelMax)
+		GLOB.channelReserved=256
+	if(!attached.attachedSounds)
+		attached.attachedSounds = list()
+	attached.attachedSounds.Add(src)
+	GLOB.allAttachedSounds.Add(src)
+
+	recalculate()
+
+/sound/proc/stop()
```
Comment (Mothblocks) [inline code/game/sound.dm:45]: All of these procs need to be documented.

https://github.com/tgstation/tgstation/blob/master/.github/AUTODOC_GUIDE.md

File: code/game/sound.dm [hunk 21]
```diff
@@ -22,6 +26,65 @@
 		0, // Flags (1 = Auto Direct, 2 = Auto Room, 4 = Auto RoomHF)
 	)
 	environment = SOUND_ENVIRONMENT_NONE //Default to none so sounds without overrides dont get reverb
+	var/atom/attached
+	var/list/listeners = list()
+	var/maxdist = 0
```
Comment (Mothblocks) [inline code/game/sound.dm:31]: Document all these.

Also, it looks like they don't clear properly if `attached` or whatever else gets deleted, meaning we're going to get hard dels.

File: code/modules/mob/mob_movement.dm (hunk 15 above)
Comment (Iamgoofball) [reply code/modules/mob/mob_movement.dm:551]: you see the [WIP] in the title?

File: code/game/sound.dm (hunk 3 above)
Comment (LemonInTheDark) [reply code/game/sound.dm:248]: According to bee this is 390k median calls and mean 500k calls per round in the last 7 days.

This is extra not an option. Not even a little bit

**Comment by Mothblocks** [comment]
We cannot afford this in our current budget.

**Pull request closed without merging**